Головнев Максим

maximka200 opened the pull request.

maximka200 committed:
- написал основные тесты
- написал основную структуру логику CircularCloudLayouter и Spiral

maximka200 committed:
- переименовал проект

maximka200 committed:
- sln файл

Dimques commented on cs/TagsCloudVisualization/Spiral.cs:
```diff
@@ -0,0 +1,92 @@
+        const double radiusStep = 1.0;  
+
+        angle += angleStep;
+        var radius = radiusStep * angle;
```
`radiusStep` - а это точно шаг? У угла шаг понятен, он прибавляется всякий раз когда точку запрашиваем, т.е. угол увеличивается на шаг. А `шаг радиуса` ничего не увеличивает, поэтому немножко запутывает.

Dimques commented on cs/TagsCloudVisualization/Spiral.cs:
```diff
@@ -0,0 +1,92 @@
+    }
+
+    private bool IntersectsWithAny(Rectangle rect) =>
+        Rectangles.Any(r => r.IntersectsWith(rect));
```
Будет очень дорого, если будет, например, много мелких квадратов. Попробуй эвристику похитрее придумать. Или посмотри в сторону геометрических алгоритмов.

Dimques commented on cs/TagsCloudVisualization/Spiral.cs:
```diff
@@ -0,0 +1,92 @@
+
+namespace TagCloud;
+
+public class Spiral(Point center)
```
Не нравится, что спираль завязана намертво на прямоугольники. Тут бы сделать одно из двух:
1. Добавить спиральке дженерик-тип с каким-нибудь `constraint`-ом, чтобы только геометрические тела можно было подавать на вхол, но у структур типа `Rectangle` маловато общих с другими типами из той же сборки методов\интерфейсов (Чтобы внутри спирали одинаково центр считать и всё в таком духе, не думая о том, каким типом спираль инициализирована - помнишь, как мы с нодами делали? У каждой ноды был метод `ConvertToHtml` и когда ты его вызываешь, неважно откуда, неважно рекурсивно или нет, то тебе не надо думать, какой именно наследник ноды ты используешь), поэтому наверное какой-то враппер нужен, мб даже вокруг коллекции структур, а не единичной штуки. Вариант муторный и сложный и не факт, что рабочий, но если интересно и будет время - можешь попробовать сделать.
2. Вынести всю логику по работе с прямоугольниками в `layouter`, а здесь оставить только ответственность за то, чтобы спиралька по заданным параметрам (кстати, можно их настраиваемыми сделать - шаги и вообще стратегию расчёта следующей точки) возвращала координату следующей точки спирали. Этот вариант попроще и вроде как подразумевается в задании.

Dimques commented on cs/TagsCloudVisualization/Spiral.cs:
```diff
@@ -0,0 +1,92 @@
+    private bool IntersectsWithAny(Rectangle rect) =>
+        Rectangles.Any(r => r.IntersectsWith(rect));
+
+    private Rectangle MoveToCenter(Rectangle rect)
```
Пока полноценную эвристику тут не вижу, но могу сказать, что просто сдвигать до упора сначала по одной оси, потом по другой может быть недостаточно - когда добавишь визуализацию попробуй комбинировать прямоугольнкии сильно разных размеров (большие и маленькие) или с большим отношением сторон. Короче говоря, эдж-кейсы проверяй тщательно.

Dimques commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,17 @@
+    public Rectangle PutNextRectangle(Size rectangleSize)
+    {
+        if (rectangleSize.Width <= 0 || rectangleSize.Height <= 0)
+            throw new ArgumentException("Размер прямоугольника должен быть положительным");
```
Ну, строго говоря, если и ширина и высота отрицательные, а под размером понимается площадь, то она положительной будет. Это скорее придирка, но давай два разных эксепшона мб кидать на каждый параметр? Либо надо придумать, как указать, какой именно параметр "плохой".

Dimques commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,17 @@
+
+namespace TagCloud;
+
+public class CircularCloudLayouter(Point center)
```
Пока непонятна ответственность этого класса.

Dimques commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,17 @@
+        if (rectangleSize.Width <= 0 || rectangleSize.Height <= 0)
+            throw new ArgumentException("Размер прямоугольника должен быть положительным");
+        
+        return RectangleSpiral.TryAddRectangleInCenter(rectangleSize, out var rectangle)
```
Как-будто многовато `CircularCloudLayouter`-у надо знать про методы спирали, не находишь? Про то, что сначала надо в центр пытаться класть и всё такое.

Dimques commented on cs/TagsCloudVisualization/Spiral.cs:
```diff
@@ -0,0 +1,92 @@
+
+    public Rectangle AddRectangle(Size rectangleSize)
+    {
+        while (true)
```
Представь, что я воткнул 4 очень больших квадрата в самом начале и все следующие прямоугольники частенько с ними пересекаются и цикл вхолостую работает. Мб научим спиральку прыгать сразу на N шагов вперёд, а стратегией выбора шагов будем управлять здесь? А-ля, с каждой _неудачей подряд_ будем увеличивать очередной шаг спирали в 2 раза (экспоненциальная стратегия) или на 5 (линейная стратегия) вместо единицы (дефолтная стратегия). А как-только получили "удачные" координаты - стратегию сбрасываем в дефолтную до следующей неудачи.
Но для этого надо уметь `быстро` сдвигать к центру!

Dimques commented on cs/TagsCloudVisualization/Spiral.cs:
```diff
@@ -0,0 +1,92 @@
+        return false;
+    }
+
+    #region Helpers
```
Выкини пока регионы, класс на ~500 строк будет норм без них смотреться, если будет больше - подумаешь, чтобы вернуть.

Dimques commented on cs/tdd.sln:
```diff
@@ -7,6 +7,10 @@ Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "BowlingGame", "BowlingGame\
 EndProject
 Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "Samples", "Samples\Samples.csproj", "{B5108E20-2ACF-4ED9-84FE-2A718050FC94}"
 EndProject
+Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "TagsCloudVisualization", "TagsCloudVisualization\TagsCloudVisualization.csproj", "{6580A616-7832-45F4-8A9A-E9A5D4FE64A8}"
```
Когда пишут библиотеку, то где-то (обычно отдельно) хранят солюшн/проект, а-ля `MyAwesomeLibraryExamples`, в котором удобно потестить новую версию с точки зрения пользовательского опыта. А ещё на такой проект можно давать ссылку пользователям - иногда это удобнее, чем ссылаться на документацию, хотя ещё удобнее, когда есть и то, и другое. Давай добавим такой же проект с примерами в наш солюшн - пусть он подключает классы из `TagsCloudVisualisation` и рисует несколько кейсов распределения прямоугольников.

maximka200 committed:
- добавил проект для теста ImageSharp

maximka200 committed:
- написал тесты
- доп. проект с примерами использования библиотеки
- визуализация упавших тестов
- переделал логику

maximka200 committed:
- поправил проект с примерами

maximka200 committed:
- дописал тестов на конфиг-билдер

Dimques commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,129 @@
+using System.Drawing;
+using ArgumentOutOfRangeException = System.ArgumentOutOfRangeException;
```
Вроде бы без этого работает

maximka200 replied:
у меня ломается почему то

> Вроде бы без этого работает

<img width="693" height="421" alt="изображение" src="https://github.com/user-attachments/assets/78b3d98e-24fb-4969-80cf-fcf33ac80877" />
<img width="606" height="302" alt="изображение" src="https://github.com/user-attachments/assets/d84ed6df-cd25-4a7c-8c5c-57d412abb983" />

maximka200 replied:
а, все понял о чем ты

Dimques commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,129 @@
+    private readonly List<Rectangle> rectangles = [];
+    public IReadOnlyCollection<Rectangle> Rectangles => rectangles.AsReadOnly();
+    
+    private int failureStreak;
```
Череда неудач - неудач чего? По названию непонятно, 1-2 слова можно еще добавить к названию.

Dimques commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,129 @@
+    private int failureStreak;
+    private int currentStep = DefaultStep;
+
+    private const int DefaultStep = 1;
```
Шаг чего?

Dimques commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,129 @@
+    public IReadOnlyCollection<Rectangle> Rectangles => rectangles.AsReadOnly();
+    
+    private int failureStreak;
+    private int currentStep = DefaultStep;
```
Попробуй управление стратегией перенести в спираль. Можно попробовать следующее - ты запрашиваешь у спирали очередную точку и если точка привела к тому, что построенный на ней прямоугольник сразу пересёкся с кем-то, значит надо уведомить спираль о неудаче и запросить следующую точку. Давай добавим спирали ещё один метод, который будет ПЕРЕзапрашивать точку (например, `RetryGetNextPoint`) - тогда спираль понимает, что если вызвали этот метод, значит, надо увеличить счётчик повторов и выбрать точку подальше на основе этого счётчика. Таким образом управление стратегией переносится в спираль и она сама решает, сколько прокрутить для очередной попытки (и сама хранит все стрики, степы и прочие дополнительные переменные про состояние).
В клауд-лэйаутере ты сначала всегда вызываешь метод GetNextPoint, а потом, если есть пересечение, в цикле дёргаешь у спирали RetryGetNextPoint, пока пересечение не уйдёт.

Dimques commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,129 @@
+        }
+    }
+
+    private Rectangle MoveAtTheWay(Rectangle rectangle, int dx, int dy, bool toCenter, int maxStep)
```
`toCenter` всегда `true`, можно не передавать и не проверять

Dimques commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,129 @@
+        var step = 1;
+        var sign = toCenter ? 1 : -1;
+
+        while (step <= maxStep)
```
Вот это странно выглядит - ты такой начинаешь с маленького шага, разгоняешься, потом тормозишь, если врезался, и возвращаешь последнюю удачную позицию, чтобы ещё разок на минимальный шаг её сдвинуть. 
Еще дельта сдвига зависит от размеров прямоугольника - мелкие помаленьку сдвигаются, большие помногу, но ведь должно быть не так, надо сдвигать помногу, если просто есть такая возможность, и наоборот - если ты рядом с облаком, сдвигать надо по чуть чуть.
Еще ты сразу сдвигаешься по обеим осям, как бы напрямую к центру, поэтому плотность хромает. А ты попробуй сначала по одной оси сдвинуться до какого-то предела, а потом по-другой - неидеальный вариант, но скорее всего плотность выше будет сразу же.

Dimques commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,129 @@
+            var before = rectangle;
+            
+            var rough = MoveAtTheWay(rectangle, dx, dy, toCenter: true, maxStep: 64);
+            
```
Вот это очень запутанное место. Во-первых, можно вот так упростить, чтобы поменьше сносило мозг (ещё лучше отказаться от переопределения переменной `rectangle`, которую дали на вход методу, но тут уже сам докручивай):
```
        while (true)
        {
            var rectCenter = GetCenter(rectangle);
            var dx = Math.Sign(center.X - rectCenter.X);
            var dy = Math.Sign(center.Y - rectCenter.Y);
            
            if (dx == 0 && dy == 0)
                return rectangle;
            
            // before больше не нужен

            var rough = MoveAtTheWay(rectangle, dx, dy, toCenter: true, maxStep: 64);
            var exact = MoveAtTheWay(rough, dx, dy, toCenter: true, maxStep: 1);
            
            if (exact == rectangle)
                return exact;
            
            rectangle = exact; // присваиваем только если надо дальше уточнять положение
        }
```
Во-вторых, не нравится цикл снаружи, который уточняет положение прямоугольника после двух сдвигов - а зачем тогда точный сдвиг-то нужен был, если получается, что сдвинулось мимо?

Dimques commented on cs/TagsCloudVisualization/Spiral.cs:
```diff
@@ -0,0 +1,21 @@
+
+public class Spiral(Point center, double angleStep = 0.01, double radiusСoeff = 1.0)
+{
+    private double angle;
```
В чём угол измеряется?

Dimques commented on cs/TagsCloudVisualization/Spiral.cs:
```diff
@@ -0,0 +1,21 @@
+        ArgumentOutOfRangeException.ThrowIfNegativeOrZero(steps);
+
+        angle += angleStep * steps;
+        var radius = radiusСoeff * angle;
```
Что за коэффициент? Точно его через параметр надо передавать?

Dimques commented on cs/TagsCloudVisualisationExamples/Examples.cs:
```diff
@@ -0,0 +1,166 @@
+
+namespace TagsCloudVisualisationExamples;
+
+[TestFixture]
```
1. Не надо в виде тестов :)
2. Ты молодец, что несколько примеров сделал, но обычно достаточно одного-двух стандартных (подумай сам, что было бы стандартным здесь, в том числе про размеры прямоугольников).

Dimques commented on cs/TagsCloudVisualization/Visualisation/CloudVisualizationConfigBuilder.cs:
```diff
@@ -0,0 +1,64 @@
+
+namespace TagCloud.Visualisation;
+
+public class CloudVisualizationConfigBuilder
```
Это лайк, однозначно, молодец!

Dimques commented on cs/TagsCloudVisualizationTests/CircularCloudLayouterTests.cs:
```diff
@@ -0,0 +1,201 @@
+        
+        var relativeFileName = Path.Combine(
+            "Failures",
+            $"{context.Test.Name}_{DateTime.Now:yyyyMMdd_HHmmss}.png");
```
До секунд маловато может быть, чтобы избежать коллизий (например, если тесты запараллелили), не стесняйся тики брать.

Dimques commented on cs/TagsCloudVisualizationTests/CircularCloudLayouterTests.cs:
```diff
@@ -0,0 +1,201 @@
+
+        var boundingWidth = maxX - minX;
+        var boundingHeight = maxY - minY;
+        var boundingArea = (double)boundingWidth * boundingHeight;
```
С такой `BoundingArea` трудно будет хорошо плотность посчитать. Алгоритм Джарвиса? :)

Dimques commented on cs/TagsCloudVisualizationTests/CircularCloudLayouterTests.cs:
```diff
@@ -0,0 +1,201 @@
+    [SetUp]
+    public void Setup()
+    {
+        center = new Point(0, 0);
```
Давай тестов на смещение центра добавим? Да, с ним должно работать :)

Dimques commented on cs/TagsCloudVisualizationTests/CloudVisualizationConfigBuilderTests.cs:
```diff
@@ -0,0 +1,131 @@
+namespace TagCloudTests;
+
+[TestFixture]
+public class CloudVisualizationConfigBuilderTests
```
Хвалю за попытку хорошо сделать, но кажется, что тесты на билдер излишни - как-будто присваивание в этих тестах проверяешь, не? :)

Dimques commented on cs/TagsCloudVisualizationTests/CloudVisualizerTests.cs:
```diff
@@ -0,0 +1,234 @@
+    }
+
+    [Test]
+    public void ConfigIsNull_ShouldThrowArgumentNullException()
```
А вот это норм тест - отсутствие конфига действительно иногда является причиной дропнуть всё приложение, это лайк.

Dimques commented on cs/TagsCloudVisualizationTests/SpiralTests.cs:
```diff
@@ -0,0 +1,41 @@
+    }
+
+    [Test]
+    public void GetNextPoint_ShouldReturnDifferentPoints()
```
Может лучше сделаем тест, который покажет, что спираль действительно раскручивается?

maximka200 committed:
- внес правки

Dimques commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,166 @@
+        return Math.Sqrt(halfW * halfW + halfH * halfH);
+    }
+    
+    private static bool CirclesIntersect(Point c1, double r1, Point c2, double r2)
```
Как-будто не хватает какого-то хэлпера, который бы считал расстояния между точками, расстояние между `X` координатами и всего в таком духе.

Dimques commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,166 @@
+            var distX = Math.Abs(center.X - shiftedCenter.X);
+            var distY = Math.Abs(center.Y - shiftedCenter.Y);
+
+            if (dx != 0 && distX >= bestDistX ||
```
В строчку норм будет смотреться

Dimques commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,166 @@
+            };
+
+            var shiftedCenter = GetCenter(shifted);
+            var distX = Math.Abs(center.X - shiftedCenter.X);
```
Во, опять хэлпер напрашивается (см.ниже).

Dimques commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,166 @@
+    private Rectangle FindBestPositionWithExponentialStep(Rectangle rectangle, int dx, int dy)
+    {
+        var best = rectangle;
+        var bestCenter = GetCenter(best);
```
Хороший кандидат на метод-расширение для `Rectangle`.

maximka200 replied:
не очень удобно его выносить, т.к. завязан на внутренних методах и некоторых полях

Dimques commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,166 @@
+    private void AddRectangle(Rectangle rect)
+    {
+        rectangles.Add(rect);
+        var c = GetCenter(rect);
```
Не экономь на буковах, от коротких слов вряд ли хуже станет читаемость.

Dimques commented on cs/TagsCloudVisualization/Spiral.cs:
```diff
@@ -0,0 +1,43 @@
+    {
+        placementFailureStreak++;
+
+        currentAngleStepMultiplier =
```
тоже в строчку норм будет

Dimques commented on cs/TagsCloudVisualizationTests/CircularCloudLayouterTests.cs:
```diff
@@ -0,0 +1,220 @@
+    [Repeat(100)]
+    public void PutNextRectangle_CloudForManyRectangles_ShouldBeDense()
+    {
+        var random = new Random(42);
```
Захардкодь тут 15 на 15 и плотность 0.8.

Dimques replied:
Ну типа, понятно, что на рандомных прямоугольниках будет ниже, но давай покажем, что в целом плотность довольно хорошо поддерживается. Либо еще можешь подобрать значения плотности и размера, которые будут у стандартных прямоугольников - высота примерно одинаковая, длина наверное 8-12, под среднее слово, короче говоря.

Dimques commented on cs/TagsCloudVisualization/Visualisation/CloudVisualizationConfigBuilder.cs:
```diff
@@ -0,0 +1,64 @@
+        return this;
+    }
+
+    public CloudVisualizationConfigBuilder WithBackground(Rgba32 color)
```
Никогда, никогда, никогда, **НИКОГДА** в публичных методах своих интерфейсов **НЕ ИСПОЛЬЗУЙ ТИПЫ ИЗ БИБЛИОТЕК-ЗАВИСИМОСТЕЙ!** Во-первых, если пользователь на них завяжется, то ты не сможешь сменить библиотеку для отрисовки под капотом у конфиг-билдера без поломки обратной совместимости с кодом пользователя, а во-вторых если он тоже использует ту же зависимость, но другой версии, могут быть очень дурацкие конфликты с зависимостями. В целом нюгет неплохо их разрешает, но всё-равно от пользователя надо стараться прятать транзитивные зависимости, для этого есть специальные инструкции в `*.csproj` файлах - если интересно, гугли про `PrivateAssets` и другие похожие инструкции.

maximka200 committed:
- поправил зависимость в конфиг билдере от внешней библиотеки цветов

maximka200 committed:
- переписал тест на кучность

maximka200 committed:
- вынес хелперы
